ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/RaceIterator.hs

```diff
@@ -0,0 +1,164 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE PatternSynonyms     #-}
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Test.Ouroboros.Consensus.ChainGenerator.RaceIterator (
+    Race (Race, UnsafeRace)
+  , RaceLbl
+  , init
+  , initConservative
+  , next
+  , nextConservative
+  ) where
+
+{- | These functions iteratively produce all race windows in a slot vector.
+
+The first window is produced by 'init', which unconditionally starts the window at the first slot.
+This window can then be passed to 'next', which starts the new window after the first active slot.
+
+@
+---X--X--X--X-- ...
+^ start of window 1 from 'init'
+    ^ start of window 2 from 'next'
+       ^ start of window 3 from 'next'
+@
+
+Valid windows must have @k+1@ active slots.
+If the vector doesn't have sufficient slots to meet this condition, 'init' and 'next' return 'Nothing' and we fall back
+to 'initConservative' and 'nextConservative', which create a window of size @scg - delta@ unconditionally, in order to
```

Comment:
I'm afraid this is still not correct. The conservative functions always clip the race windows to the length of the containing sequence, so in practice they don't extend beyond the end.

Reply:
Addressed in 5e0cc0797